### PR TITLE
done

### PR DESCRIPTION
**Pull request by necsanuradu**
complete, more testing maybe

### CODE REVIEW

File: spec/SpecHelper.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+beforeEach(function () {
```
Comment (edpackard) [inline spec/SpecHelper.js:1]: could delete this file?

File: .vscode/.prettierrc.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (edpackard) [inline .vscode/.prettierrc.json:1]: should this file be in a git ignore file?

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (edpackard) [inline .vscode/settings.json:1]: should this file be in a git ignore file?

File: Gemfile.lock [hunk 4]
```diff
@@ -0,0 +1,152 @@
+GEM
```
Comment (edpackard) [inline Gemfile.lock:1]: can add this file to git ignore file

File: SpecRunner.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    <!-- include source files here... -->
+    <script src="src/frame.js"></script>
+    <script src="src/game.js"></script>
+    <script src="src/loadingFrames.js"></script>
+    <script src="src/framesHtml.js"></script>
+
+    <!-- include spec files here... -->
+    <script src="spec/SpecHelper.js"></script>
+    <!--<script src="spec/PlayerSpec.js"></script>-->
```
Comment (edpackard) [inline SpecRunner.html:28]: could delete this line?

File: controller/app.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+require 'sinatra'
```
Comment (edpackard) [inline controller/app.rb:1]: I like how you've used sinatra to serve the html

File: controller/app.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+class BowlingGame < Sinatra::Base
+
+  configure :development do
+    register Sinatra::Reloader
+  end
+
+  enable :sessions
+  
+  get '/' do  
+    p 'index.html'.to_sym
```
Comment (edpackard) [inline controller/app.rb:13]: is this line necessary?

File: controller/views/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,99 @@
+<!DOCTYPE html>
```
Comment (edpackard) [inline controller/views/index.html.erb:1]: wow! there's a lot of work here - I'm tempted to clone your repo and see what it looks like :)

File: spec/jasmine/frameSpec.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    it("a new frame should have the frame bonus 0", function () {
+      expect(frame.points("bonus")).toEqual(0);
+    });
+
+    it("a new frame should have the total score 0", function () {
+      expect(frame.points()).toEqual(0);
+    });
+
+    it("a new frame should have the total score 3", function () {
```
Comment (edpackard) [inline spec/jasmine/frameSpec.js:20]: maybe clearer 'it' message 'a roll should increase the frame score' ?

File: spec/jasmine/frameSpec.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    });
+
+    it("frame after adding bonus and score should have the total 7", function () {
+      frame._addRollAndReturnBonusForPrevFrames(3);
+      frame.addBonusSentFromFutureRolls(4);
+      expect(frame.points()).toEqual(7);
+    });
+  });
+
+  describe("test _addRollAndReturnBonusForPrevFrames return bonusPrevFrames", function () {
```
Comment (edpackard) [inline spec/jasmine/frameSpec.js:37]: write test descriptions in plain English as much as possible so the Jasmine output reads nicely

File: spec/jasmine/frameSpec.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+  it("should throw error is we pass to addRoll.. but the frame hase ended", function () {
+    expect(function () {
+      frame._addRollAndReturnBonusForPrevFrames(10);
+      frame._addRollAndReturnBonusForPrevFrames(1);
+    }).toThrow("this frame has eneded");
+  });
+});
+
+//expect(fizz.run(1)).toBeTruthy();
```
Comment (edpackard) [inline spec/jasmine/frameSpec.js:130]: delete these comments?

File: src/frame.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      return true;
+    else return false;
+  }
+
+  tooManyPins(pins) {
+    if (this.frameScore.reduce((a, b) => a + b, 0) + pins > 10) return true;
+    else return false;
+  }
+
+  points(what) {
```
Comment (edpackard) [inline src/frame.js:67]: maybe `type` or something more descriptive instead of `what`

File: src/frame.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
+class FrameNotTen {
```
Comment (edpackard) [inline src/frame.js:1]: could you use more private variables/functions in this class for encapsulation purposes?

File: src/frame.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  checkForErrors(pins) {
+    if (isNaN(pins)) this.shouldThrowError("argument should be a number");
+    if (pins < 0) this.shouldThrowError("argument should be greater than 0");
+    if (this.frameEnded()) this.shouldThrowError("this frame has eneded");
+    if (this.tooManyPins(pins) && this.frameNumber != 10)
+      this.shouldThrowError("argument + score > 10");
+  }
+
+  frameEnded() {
+    if (this.frameNumber != 10 && this.frameScore.length >= 2) return true;
```
Comment (edpackard) [inline src/frame.js:39]: clear logic tree here, could it be even clearer if broken down into component functions?

File: src/framesHtml.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
+class FramesHtml {
+  _doNumber(frame, fEl) {
```
Comment (edpackard) [inline src/framesHtml.js:2]: this class looks really interesting - is it generating html output using JS?

**Review by edpackard — COMMENTED** [review]
I'm impressed by how much work has gone into this, and that you've really engaged with JSON, html and css. I'm really tempted to clone your repo and see how it looks! I like that you've used sinatra so that I could presumably view your app hosted locally on my computer.

On that note, a readme would be really useful, explaining how to install and use the program. Without usage instructions, I'm not entirely sure how I would interact with the program - this makes it trickier to figure out the flow of your code and how the classes interact. So my main tip: always provide a readme with basic instructions (and paste a screenshot of the final product, especially if it looks cool).

In terms of testing, you possibly could use more spies to mock classes that aren't the class you are testing.
When testing, write slightly clearer (non-'codey') description messages so that the Jasmine output has an easy-to-follow narrative. Similarly, in your actual code, do think about the order in which you place your functions to try and make the class look as much like a 'narrative' as possible. Clear, logical variable and function names will also help a reader follow your logic.

Otherwise, this all looks very impressive in terms of your grasp of JavaScript and how it interacts with html - I was a little lost in these areas at times, most likely because I'm not as experienced in this.

**Comment by necsanuradu** [comment]
Thank you Ed for checking my code. You are absolutely right on all the issues you raised specially on the README file. I've made some mistakes which costed me extra time and reduced my willingness to even look at my code with good eyes other that to make sure it produced the expected results. I've also missed my chance to make good use of the Jasmine spies to stop the chain of execution towards DOM manipulation and that resulted into not testing the total score other than on the browser. Overall I underestimated Javascript's complicated ways of execute and return. Thank you again.
